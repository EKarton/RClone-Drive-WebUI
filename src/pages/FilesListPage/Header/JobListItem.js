import ListItem from '@mui/material/ListItem';
import ListItemText from '@mui/material/ListItemText';
import { useEffect, useState } from 'react';
import { JobStatus } from 'services/RCloneJobTracker/constants';
import { getFullPath } from 'utils/filename-utils';

export default function JobListItem({ job }) {
  const [status, setStatus] = useState(null);

  useEffect(() => {
    const subscriber = job.status.subscribe((newStatus) => {
      setStatus(newStatus);
    });

    return () => {
      subscriber.unsubscribe();
    };
  });

  const getMessage = () => {
    switch (job.jobType) {
      case 'MOVE_FILE':
      case 'MOVE_FOLDER': {
        return getMoveMessage();
      }
      case 'RENAME_FILE':
      case 'RENAME_FOLDER': {
        return getRenameMessage();
      }
      case 'UPLOAD_FILE': {
        return getUploadMessage();
      }
      default: {
        return getDefaultString();
      }
    }
  };

  const getMoveMessage = () => {
    const { src, target } = job;
    const srcStr = getFileString(src);
    const targetStr = getFileString(target);

    if (status === JobStatus.ERROR) {
      return `Failed to move ${srcStr} to ${targetStr}`;
    }

    if (status === JobStatus.SUCCESS) {
      return `Moved ${srcStr} to ${targetStr}`;
    }

    return `Moving ${srcStr} to ${targetStr}`;
  };

  const getRenameMessage = () => {
    const { src, target } = job;
    const remotePath = `${src.remote}:${src.dirPath}`;

    if (status === JobStatus.ERROR) {
      return `Failed to rename ${src.name} to ${target.name} in ${remotePath}`;
    }

    if (status === JobStatus.SUCCESS) {
      return `Renamed ${src.name} to ${target.name} in ${remotePath}`;
    }

    return `Renaming ${src.name} to ${target.name} in ${remotePath}`;
  };

  const getUploadMessage = () => {
    const { fileName } = job;

    if (status === JobStatus.ERROR) {
      return `Failed to upload ${fileName}`;
    }

    if (status === JobStatus.SUCCESS) {
      return `Uploaded ${fileName}`;
    }

    return `Uploading ${fileName}`;
  };

  const getDefaultString = () => {
    if (status === JobStatus.ERROR) {
      return 'Failed job';
    }

    if (status === JobStatus.SUCCESS) {
      return 'Job succeeded';
    }

    return 'Job running';
  };

  const getFileString = (file) => {
    return getFullRemotePath(file.remote, getFullPath(file.dirPath, file.name));
  };

  const getFullRemotePath = (remote, fullPath) => {
    return `${remote}:${fullPath}`;
  };

  return (
    <ListItem>
      <ListItemText primary={getMessage()} />
    </ListItem>
  );
}
