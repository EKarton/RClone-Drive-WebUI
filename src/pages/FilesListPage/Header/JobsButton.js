import CloudDoneIcon from '@mui/icons-material/CloudDone';
import Button from '@mui/material/Button';
import CircularProgress from '@mui/material/CircularProgress';

export default function JobsButton({ statusCounts, ...props }) {
  const { numOngoing, numSuccessful, numFailed } = statusCounts;

  if (numOngoing === 0 && numSuccessful === 0 && numFailed === 0) {
    return null;
  }

  if (numOngoing > 0) {
    return (
      <Button startIcon={<CircularProgress size={20} />} {...props}>
        {numOngoing} Jobs Running
      </Button>
    );
  }

  return (
    <Button startIcon={<CloudDoneIcon />} {...props}>
      {numSuccessful} Jobs Finished
    </Button>
  );
}
