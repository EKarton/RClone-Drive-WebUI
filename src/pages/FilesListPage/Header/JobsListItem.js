import ListItem from '@mui/material/ListItem';
import ListItemText from '@mui/material/ListItemText';
import { useEffect, useState } from 'react';
import { JobStatus, JobTypes } from 'utils/constants';
import { getFullPath } from 'utils/filename-utils';

export default function JobsListItem({ job }) {
  const [status, setStatus] = useState(null);

  useEffect(() => {
    const subscriber = job.status.subscribe((newStatus) => {
      setStatus(newStatus);
    });

    return () => {
      subscriber.unsubscribe();
    };
  });

  const getMessage = () => {
    switch (job.jobType) {
      case JobTypes.MOVE_FILE:
      case JobTypes.MOVE_FOLDER: {
        return getMoveMessage();
      }
      case JobTypes.RENAME_FILE:
      case JobTypes.RENAME_FOLDER: {
        return getRenameMessage();
      }
      case JobTypes.UPLOAD_FILE: {
        return getUploadMessage();
      }
      default: {
        return getDefaultString();
      }
    }
  };

  const getMoveMessage = () => {
    const { src } = job;
    const srcStr = getFileString(src);

    if (status === JobStatus.ERROR) {
      return `Failed to move ${srcStr}`;
    }

    if (status === JobStatus.SUCCESS) {
      return `Moved ${srcStr}`;
    }

    return `Moving ${srcStr}`;
  };

  const getRenameMessage = () => {
    const { src, target } = job;

    if (status === JobStatus.ERROR) {
      return `Failed to rename ${src.name}`;
    }

    if (status === JobStatus.SUCCESS) {
      return `Renamed ${src.name} to ${target.name}`;
    }

    return `Renaming ${src.name} to ${target.name}`;
  };

  const getUploadMessage = () => {
    const { name } = job;

    if (status === JobStatus.ERROR) {
      return `Failed to upload ${name}`;
    }

    if (status === JobStatus.SUCCESS) {
      return `Uploaded ${name}`;
    }

    return `Uploading ${name}`;
  };

  const getDefaultString = () => {
    if (status === JobStatus.ERROR) {
      return 'Failed job';
    }

    if (status === JobStatus.SUCCESS) {
      return 'Job succeeded';
    }

    return 'Job running';
  };

  const getFileString = (file) => {
    return getFullRemotePath(file.remote, getFullPath(file.dirPath, file.name));
  };

  const getFullRemotePath = (remote, fullPath) => {
    return `${remote}:${fullPath}`;
  };

  return (
    <ListItem>
      <ListItemText primary={getMessage()} />
    </ListItem>
  );
}
