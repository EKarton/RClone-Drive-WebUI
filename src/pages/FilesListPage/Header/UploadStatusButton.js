import CloudDoneIcon from '@mui/icons-material/CloudDone';
import Button from '@mui/material/Button';
import CircularProgress from '@mui/material/CircularProgress';
import { useFileUploadCounts } from 'contexts/FileUploadCounts';
import { useFileUploadDialog } from 'contexts/FileUploadDialog/index';

export default function UploadStatusButton() {
  const { openDialog } = useFileUploadDialog();
  const { counts } = useFileUploadCounts();
  const { numUploading, numSuccessful } = counts;

  if (numUploading === 0 && numSuccessful === 0) {
    return null;
  }

  if (numUploading === 0) {
    return (
      <Button startIcon={<CloudDoneIcon />} onClick={openDialog}>
        {numSuccessful} Uploaded
      </Button>
    );
  }

  return (
    <Button startIcon={<CircularProgress size={20} />} onClick={openDialog}>
      Uploading {numUploading} Files
    </Button>
  );
}
