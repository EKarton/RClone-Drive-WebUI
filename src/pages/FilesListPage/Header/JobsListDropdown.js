import List from '@mui/material/List';
import Popover from '@mui/material/Popover';
import JobListItem from './JobListItem';

export default function JobsListDropdown({ open, jobs, anchorEl, onClose }) {
  const recentJobs = jobs.slice(Math.max(0, jobs.length - 4), jobs.length).reverse();

  return (
    <Popover
      open={open}
      anchorEl={anchorEl}
      onClose={onClose}
      anchorOrigin={{
        vertical: 'bottom',
        horizontal: 'right',
      }}
      transformOrigin={{
        vertical: 'top',
        horizontal: 'right',
      }}
    >
      <List dense>
        {recentJobs.map((job) => (
          <JobListItem job={job} />
        ))}
      </List>
    </Popover>
  );
}
