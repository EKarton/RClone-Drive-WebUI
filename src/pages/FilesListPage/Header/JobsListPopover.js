import Divider from '@mui/material/Divider';
import List from '@mui/material/List';
import ListItemButton from '@mui/material/ListItemButton';
import ListItemText from '@mui/material/ListItemText';
import Popover from '@mui/material/Popover';
import JobsListItem from './JobsListItem';

export default function JobsListPopover(props) {
  const { open, jobs, anchorEl, onClose, onMoreClicked } = props;
  const recentJobs = jobs.slice(Math.max(0, jobs.length - 4), jobs.length);

  return (
    <Popover
      open={open}
      anchorEl={anchorEl}
      onClose={onClose}
      anchorOrigin={{
        vertical: 'bottom',
        horizontal: 'right',
      }}
      transformOrigin={{
        vertical: 'top',
        horizontal: 'right',
      }}
    >
      <List dense>
        {recentJobs.map((job) => (
          <JobsListItem key={job.jobId} job={job} />
        ))}
        <Divider />
        <ListItemButton onClick={onMoreClicked} data-testid="more-details-button">
          <ListItemText primary="More details" />
        </ListItemButton>
      </List>
    </Popover>
  );
}
