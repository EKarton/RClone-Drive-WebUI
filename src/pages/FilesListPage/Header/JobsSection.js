import { useState } from 'react';
import { useJobsListDialog } from 'contexts/JobsListDialog';
import useJobQueueInfo from 'hooks/jobs/useJobQueueInfo';
import JobsButton from './JobsButton';
import JobsListPopover from './JobsListPopover';

export default function JobsSection() {
  const { show } = useJobsListDialog();
  const { jobs, statusCounts } = useJobQueueInfo();
  const [anchorEl, setAnchorEl] = useState(null);

  const handleClick = (e) => {
    setAnchorEl(e.currentTarget);
  };

  const handleClose = () => {
    setAnchorEl(null);
  };

  const handleMoreButtonClicked = () => {
    setAnchorEl(null);
    show();
  };

  return (
    <>
      <JobsButton
        statusCounts={statusCounts}
        onClick={handleClick}
        data-testid="job-button"
      />
      <JobsListPopover
        open={anchorEl !== null}
        jobs={jobs}
        anchorEl={anchorEl}
        onClose={handleClose}
        onMoreClicked={handleMoreButtonClicked}
      />
    </>
  );
}
