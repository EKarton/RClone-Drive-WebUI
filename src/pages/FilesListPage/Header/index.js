import Breadcrumbs from 'components/Breadcrumbs';
import JobsSection from './JobsSection';
import './index.scss';

export default function Header({ remote, path }) {
  return (
    <div className="header">
      <Breadcrumbs remote={remote} path={path} homeText="My Files" homePath="/files" />
      <JobsSection />
    </div>
  );
}
