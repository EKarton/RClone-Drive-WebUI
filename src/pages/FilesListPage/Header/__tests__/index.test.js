import { customRender } from 'test-utils/react';
import Header from '../index';

describe('Header', () => {
  it('should render the header and jobs section given no jobs and mock remote and path', () => {
    const { baseElement } = customRender(<Header remote="gdrive" path="Documents" />);

    expect(baseElement).toMatchSnapshot();
  });
});
