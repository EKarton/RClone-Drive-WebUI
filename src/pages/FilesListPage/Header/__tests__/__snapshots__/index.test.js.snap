// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should render the header and jobs section given no jobs and mock remote and path 1`] = `
<body>
  <div>
    <div
      class="header"
    >
      <nav
        aria-label="breadcrumb"
        class="MuiTypography-root MuiTypography-body1 MuiBreadcrumbs-root css-63vwl1-MuiTypography-root-MuiBreadcrumbs-root"
      >
        <ol
          class="MuiBreadcrumbs-ol css-4pdmu4-MuiBreadcrumbs-ol"
        >
          <li
            class="MuiBreadcrumbs-li"
          >
            <a
              class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1kgfegd-MuiTypography-root-MuiLink-root"
              href="/files"
            >
              My Files
            </a>
          </li>
          <li
            aria-hidden="true"
            class="MuiBreadcrumbs-separator css-1wuw8dw-MuiBreadcrumbs-separator"
          >
            /
          </li>
          <li
            class="MuiBreadcrumbs-li"
          >
            <a
              class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1kgfegd-MuiTypography-root-MuiLink-root"
              href="/files/Z2RyaXZlOg=="
            >
              gdrive
            </a>
          </li>
          <li
            aria-hidden="true"
            class="MuiBreadcrumbs-separator css-1wuw8dw-MuiBreadcrumbs-separator"
          >
            /
          </li>
          <li
            class="MuiBreadcrumbs-li"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-1gqyz35-MuiTypography-root"
            >
              Documents
            </p>
          </li>
        </ol>
      </nav>
    </div>
  </div>
</body>
`;
